FIP-0095: First draft of add FEVM precompile to fetch beacon digest from chain history (#1048)
filecoin-project/FIPs: FIPS/fip-rand-precompile.md

This proposal provides that hook-up.

The EVM does already provide a PREVRANDAO opcode, but this is limited to access randomness from the immediately prior epoch.
This is primitive is not suited to protocols where the randomness is incorporated into off-chain protocols, 
since the prover must observe the randomness, use it for computation, and submit a message that must be included in the immediate next epoch 
in order for a smart contract to be able to verify that the correct randomness source was used.
Access to values from prior epochs allows for protocols that commit to a challenge before it is used,
take more than one epoch to complete, and/or are robust to short-lived chain instability or forks.

The concrete use case motivating the authors is to make it possible to build a proof-of-data possession in user-deployed smart contracts

## Specification
A new precompile is added to the built-in EVM actor, which implements all user-deployed smart contracts.

The precompile is callable at address `0xfe00..04`.

| Param            | Value                     |
|------------------|---------------------------|
| randomness_epoch | U256 - low i64             |

> The call errors with `PrecompileError::InvalidInput` if the epoch is invalid.

Reviewer: should outline what "invalid" means here, just future epochs? negative i64's?

Reviewer 2: I'v adjusted it slightly, but I think it's easier to define this via what positively is valid than the inverse.

The existing syscall has a gas cost function that scales linearly with the number of epochs that have elapsed since 
the requested epoch.

## Design Rationale
The beacon value is provided in as raw a form as possible, given the FVM already hashes the beacon randomness.
Including further entropy is left to the caller.

Adding a new syscall to return the un-hashed beacon randomness would add complexity without adding fundamental capability.
If a full drand beacon is required (e.g. for time-lock operations), a user-submitted beacon can be validated by hashing it
and comparing with the value returned by the new precompile.

## Backwards Compatibility
This proposal introduces a new precompile available to FEVM smart contracts.

Any existing smart contract which invokes this previously-unbound address will now observe a successful call.

## Test Cases
As part of implementation we should have:
1. Builtin actors tests that the precompile is accessible and returning the correct randomness as specified by the testing runtime.  
2. Integration tests on a lotus devent with a simple solidity contract that uses the precompile.
3. Integration tests that validate that gas costs are as expected

> 4. Integration tests that validate that very deep lookback has computational cost in proportion to the gas cost (linear).

Reviewer: ```suggestion
4. Integration tests that validate that very deep lookback has computational cost in proportion to the gas cost (linear).
5. Integration tests that validate that requesting future randomness results in an erorr.
```

---

> fip: "<to be assigned>" <!--keep the qoutes around the fip number, i.e: `fip: "0001"`-->

Reviewer 3: please t ake 0095

title: Add EVM precompile to fetch beacon digest from chain history 
author: @ZenGround0, @anorth
discussions-to: https://github.com/filecoin-project/FIPs/discussions/1051
status: Draft
type: Technical Core
created: 2024-08-26
---

> # FIP-00XX: Add EVM precompile to fetch beacon digest from chain history 

Reviewer 3: hrm isn't this a [Filecoin precompile](https://github.com/filecoin-project/FIPs/blob/de7ccac837d576a440f6210f393c1e4c98959dbf/FIPS/fip-0054.md?plain=1#L669)? If we follow the FVM spec (previous FIPs),  EVM precompile may be [mistaken as Ethereum precompiles ](https://github.com/filecoin-project/FIPs/blob/de7ccac837d576a440f6210f393c1e4c98959dbf/FIPS/fip-0054.md?plain=1#L650) imho.

Reviewer 2: That terminology is really bad. Filecoin doesn't have precompiles. The FVM has syscalls. Only the EVM actor has a concept of precompile, because it's an EVM thing. There's no difference between "EVM precompiles" and "Filecoin precompiles" except that some of them are also available in other EVMs outside Filecoin.

I'll change it to "FEVM precompile" though.

Reviewer 3: Okay with your counter proposal

## Simple Summary

> Adds a precompile to the EVM actor to allow smart contracts to fetch drand beacon digests from the chain history.

Reviewer 3: ```suggestion
Adds a precompile to the EVM actor to allow EVM smart contracts to fetch drand beacon digests from the chain history.
```

how far back can we support? since genesis?

Reviewer 2: > how far back can we support? since genesis?

Yes

## Abstract
The FVM exposes a syscall for fetching the digest of a drand beacon included in some prior Filecoin chain header.
The built-in actors use this facility to provide randomness for storage proofs.
This proposal introduces an EVM actor precompile exposing this functionality to user-deployed smart contracts
so that they can also use random challenges for other types of proofs.

## Change Motivation
A source of un-biasable randomness is a useful primitive in proofs of service, 
as evidenced by its use in Filecoin's existing proof of storage protocol.
The Filecoin blockchain already embodies all the infrastructure to provide this primitive, 
but is missing the hook-up to where user-deployed smart contracts can use it. 
This proposal provides that hook-up.

The EVM does already provide a PREVRANDAO opcode, but this is limited to access randomness from the immediately prior epoch.

> This is primitive is not suited to protocols where the randomness is incorporated into off-chain protocols, 

Reviewer 3: ```suggestion
This primitive is not suited to protocols where the randomness is incorporated into off-chain protocols, 
```

since the prover must observe the randomness, use it for computation, and submit a message that must be included in the immediate next epoch 
in order for a smart contract to be able to verify that the correct randomness source was used.
Access to values from prior epochs allows for protocols that commit to a challenge before it is used,
take more than one epoch to complete, and/or are robust to short-lived chain instability or forks.

The concrete use case motivating the authors is to make it possible to build a proof-of-data possession in user-deployed smart contracts

## Specification
A new precompile is added to the built-in EVM actor, which implements all user-deployed smart contracts.

> The precompile is callable at address `0xfe00..06`.

Reviewer 3: ```suggestion
The precompile is callable at Ethereum address `0xfe00..06`.
```

| Param            | Value                     |
|------------------|---------------------------|

> | epoch            | U256 - low i64            |

Reviewer 3: ```suggestion
| epoch            | u256 - low i64            |
```

Reviewer 3: is the return value the beacon round or the beacon entry (array)?

Reviewer 2: It's the hash of the randomness value from the beacon entry (described below)

The call errors with `PrecompileError::InvalidInput` if the epoch is not a valid, prior epoch number.

> This calls directly to the FVMs `get_beacon_randomness` syscall, which returns a Blake2b hash digest of the 

Reviewer 3: ```suggestion
This calls directly to the FVM's `get_beacon_randomness` syscall, which returns a Blake2b hash digest of the 
```

> randomness value from the drand beacon at the associated Filecoin epoch.

Reviewer 3: what would happen if it was a null round of a given filecoin epoch?

Reviewer 2: Whatever would already happen in that case, which this FIP doesn't propose to change.

I can see a reasonable motivation to want to describe this for potential users, which now extend beyond built-in actors. This FIP isn't proposing to change any of the semantics, but they are poorly documented at present. I will try to come back with some edits to describe them here as context, but I need some help because I wasn't involved in developing the underlying syscall.

---
fip: "<to be assigned>" <!--keep the qoutes around the fip number, i.e: `fip: "0001"`-->

> title: get_randomness filecoin precompile for EVM actor 

Reviewer 3: ```suggestion
title: Add get_randomness to Filecoin-specific precompile for EVM smart contracts 
```

Reviewer 3: (mainly following the [language in the previous FIPs ](https://github.com/filecoin-project/FIPs/blob/3422557f77590508fd8f1dbabbc58fadd15a1bca/FIPS/fip-0054.md?plain=1#L671) so it is easy to search, at least in this repo

author: @anorth, @ZenGround0
discussions-to: https://github.com/filecoin-project/FIPs/discussions/1051
status: Draft
type: Technical Core
created: 2024-08-26
---

<!--You can leave these HTML comments in your merged FIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new FIPs. Note that a FIP number will be assigned by an editor. When opening a pull request to submit your FIP, please use an abbreviated title in the filename, `fip-draft_title_abbrev.md`. The title should be 44 characters or less.-->

# FIP-00XX: get_randomness filecoin precompile for EVM actor 

## Simple Summary
<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the FIP.-->

> Add a filecoin precompile to the EVM actor to allow user actors to read randomness from the drand beacon at a specified epoch.

Reviewer 3: ```suggestion
Add a Filecoin precompile to the EVM actor to allow EVM smart contracts to read randomness from the drand beacon at a specified epoch.
```

please reject this suggestion if these precompiles can be called by wasm actors in the future as well. cc @Stebalien for feedback

## Abstract
<!--A short (~200 word) description of the technical issue being addressed.-->

> The FVM exposes a syscall for reading randomness from the drand beacon at a specified epoch. The core filecoin protocol uses this randomness facility to validate proofs.  This FIP introduces a precompile exposing this FVM functionality to the fevm builtin actor allowing user contracts to make use of the same functionality.

Reviewer 3: ```suggestion
The FVM exposes a syscall for reading randomness from the drand beacon at a specified epoch. The core Filecoin protocol uses this randomness facility to validate proofs.  This FIP introduces a precompile exposing this FVM functionality to the evm builtin actor allowing user contracts to make use of the same functionality.
```

## Change Motivation
<!--The motivation is critical for FIPs that want to change the Filecoin protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the FIP solves. FIP submissions without sufficient motivation may be rejected outright.-->

> In order to build Proof of Data Possession we need a good source of randomness with a lookback. It is likely that many other applications have similar needs.

Reviewer 3: If possible, I would love a bit more elaboration on the use cases/why where apps need randomness with _a look back_

## Specification
<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Filecoin implementations. -->

> This precompile was previously implemented before the FeVM upgrade but was removed here: https://github.com/filecoin-project/builtin-actors/pull/995/files.  

Reviewer 3: i don't think this sentence/reference contributes to the specification of this FIP too much, I would suggest remove this; (if it was referred, i'd be interested in knowing why it was excluded from the introduction of FEVM

We propose a simpler variant of the previous implementation.

> ```rust

Reviewer 3: ```suggestion
```solidity
```

(small nit)

Reviewer 2: This is defined in Rust.

/// Params:
///
/// | Param            | Value                     |
/// |------------------|---------------------------|
/// | randomness_epoch | U256 - low i64             |
/// | entropy_length   | U256 - low u32             |
/// | entropy          | input\[32..] (right padded)|
///
/// any bytes in between values are ignored
///

> /// Returns empty array if invalid randomness type

Reviewer 3: can we be more specific with the return data layout?

> /// Errors if unable to fetch randomness

Reviewer 3: remove comment `///`

```

This will be defined at precompile address `0xfe00..06` and named get_randomness

## Design Rationale
<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->

> Since an almost complete design and (abandoned) implementation already exist  a precompile calling the fvm syscall was the obvious choice to get per-epcoh randomness exposed to user actors.

Reviewer 3: this paragraph only makes me wonder why it was dropped from the introduction of FEVM in the first place, was there any design and/or security concern back then?

> We are dropping the personalization tag and the randomness type parameter from the interface.  With respect to the randomness type param for our motivating application we only need one source of randomness and the unbiasable beacon is the best choice.  If other applications want VRF randomness it makes sense to leave this for another precompile and another FIP.

Reviewer 3: >   for our motivating application 

how about other applications/ use cases? would this design be suitable for most use cases?

Reviewer 3: (i think the following paragraph may capture the answer to this q - please consider moving it inline here

The existing syscall has a gas cost function that scales linearly with the number of epochs that have elapsed since 
the requested epoch.

## Design Rationale
The beacon value is provided in as raw a form as possible, given the FVM already hashes the beacon randomness.
Including further entropy is left to the caller.

Adding a new syscall to return the un-hashed beacon randomness would add complexity without adding fundamental capability.
If a full drand beacon is required (e.g. for time-lock operations), a user-submitted beacon can be validated by hashing it
and comparing with the value returned by the new precompile.

## Backwards Compatibility
This proposal introduces a new precompile available to FEVM smart contracts.

Any existing smart contract which invokes this previously-unbound address will now observe a successful call.

## Test Cases
As part of implementation we should have:
1. Builtin actors tests that the precompile is accessible and returning the correct randomness as specified by the testing runtime.  
2. Integration tests on a lotus devent with a simple solidity contract that uses the precompile.
3. Integration tests that validate that gas costs are as expected

> 4. Integration tests that validate that very deep lookback has computational cost in proportion to the gas cost (linear).

Author: I'm mildly worried about hitting edge cases in the client that are more correctness issues than cost issues.  Its unlikely but we are potentially hitting very dormant code paths.  It would be good to understand when the `LimitExceeded` error comes into play during FVM querying so we can bound our concern.

Reviewer 4: I believe that part of the FVM docs is simply out of date. https://github.com/filecoin-project/ref-fvm/pull/2034
